line_profiler/_line_profiler.pyx

```diff
@@ -164,72 +264,129 @@ cdef class LineProfiler:
+                if code_hash in c_last_time:
+                    py_last_time[code] = c_last_time[code_hash]
+        return py_last_time
+
+
+    cpdef disable(self):
+        self._c_last_time[threading.get_ident()].clear()
         unset_trace()
 
     def get_stats(self):
         """ Return a LineStats object containing the timings.
         """
+        cdef dict cmap
+        
         stats = {}
-        for code in self.code_map:
-            entries = self.code_map[code].values()
+        for code in self.code_hash_map:
+            cmap = self._c_code_map
+            entries = []
+            for entry in self.code_hash_map[code]:
+                entries += list(cmap[entry].values())
             key = label(code)
-            stats[key] = [e.astuple() for e in entries]
-            stats[key].sort()
-        return LineStats(stats, self.timer_unit)
-
-
-cdef class LastTime:
-    """ Record the last callback call for a given line.
-    """
-    cdef int f_lineno
-    cdef PY_LONG_LONG time
-
-    def __cinit__(self, int f_lineno, PY_LONG_LONG time):
-        self.f_lineno = f_lineno
-        self.time = time
 
+            entries = [(e["lineno"], e["nhits"], e["total_time"]) for e in entries]
+            # If there are multiple entries for a line, this will sort them by increasing # hits
+            entries.sort()
+
+            # NOTE: v4.x may produce more than one entry per line. For example:
+            #   1:  for x in range(10):
+            #   2:      pass
+            #  will produce a 1-hit entry on line 1, and 10-hit entries on lines 1 and 2
+            #  This doesn't affect `print_stats`, because it uses the last entry for a given line (line number is
+            #  used a dict key so earlier entries are overwritten), but to keep compatability with other tools,
+            #  let's only keep the last entry for each line
+            # Remove all but the last entry for each line
+            entries = list({e[0]: e for e in entries}.values())
+            stats[key] = entries
+        return LineStats(stats, self.timer_unit)
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
 cdef int python_trace_callback(object self_, PyFrameObject *py_frame, int what,
-    PyObject *arg):
+PyObject *arg):
     """ The PyEval_SetTrace() callback.
     """
     cdef LineProfiler self
-    cdef object code, key
-    cdef dict line_entries, last_time
-    cdef LineTiming entry
+    cdef object code
+    cdef LineTime entry
     cdef LastTime old
+    cdef int key
     cdef PY_LONG_LONG time
+    cdef int64 code_hash
+    cdef int64 block_hash
+    cdef unordered_map[int64, LineTime] line_entries
 
     self = <LineProfiler>self_
-    last_time = self.last_time
 
     if what == PyTrace_LINE or what == PyTrace_RETURN:
-        code = <object>py_frame.f_code
-        if code in self.code_map:
+        block_hash = hash(get_frame_code(py_frame))
+        code_hash = compute_line_hash(block_hash, py_frame.f_lineno)
```

Comment:
Is there a memory leak here? The `get_frame_code` function calls PyFrame_GetCode, which returns a `PyCodeObject*`, and that has its reference count decreased in the same function by `Py_DECREF`. However, the `PyObject*` returned by `PyCode_GetCode` is returned from that function, and (correct me if I'm wrong) I believe it returns a "strong reference" meaning that we need to manually call `DECREF`, but it never has it's reference count decreased later on. Is that an oversight, or am I misunderstanding `PyCode_GetCode`?

Reply:
Your understanding is absolutely correct. However, the key piece of this, which I should add a comment mentioning, is that Cython takes care of the DECREF for us. If you compile with the annotated HTML, it shows a bit of yellow at the hash, due to the DECREF. When I get to my computer, I'll add a comment in explaining that.